posthog/models/person_overrides/sql.py

```diff
@@ -0,0 +1,115 @@
+# Within PostHog, it is possible that two Persons are merged together. The
+# impact of this is that all events that were associated with the Persons should
+# now appear to be associated with a single Person.
+#
+# In the ClickHouse `sharded_events` table we have a `person_id` column that
+# contains the UUID of the Person that the event is associated with. When a
+# merge happens, we do not immediately update the `person_id` column in the
+# `sharded_events` table. Instead, we create a new row in the `person_overrides`
+# table that contains the mapping from the `old_person_id` to the
+# `override_person_id`. This allows us to OUTER JOIN the `person_overrides`
+# table to the `sharded_events` table to find all events that were associated
+# and therefore reconcile the events to be associated with the same Person.
+
+from posthog.kafka_client.topics import KAFKA_PERSON_OVERRIDE
+from posthog.settings.data_stores import CLICKHOUSE_CLUSTER, CLICKHOUSE_DATABASE, KAFKA_HOSTS
+
+PERSON_OVERRIDES_CREATE_TABLE_SQL = f"""
+    CREATE TABLE IF NOT EXISTS `{CLICKHOUSE_DATABASE}`.`person_overrides`
+    ON CLUSTER '{CLICKHOUSE_CLUSTER}' (
+        team_id INT NOT NULL,
+
+        -- When we merge two people `old_person_id` and `override_person_id`, we
+        -- want to keep track of a mapping from the `old_person_id` to the
+        -- `override_person_id`. This allows us to join with the
+        -- `sharded_events` table to find all events that were associated with
+        -- the `old_person_id` and update them to be associated with the
+        -- `override_person_id`.
+        old_person_id UUID NOT NULL,
+        override_person_id UUID NOT NULL,
+
+        -- The timestamp the merge of the two people was completed.
+        merged_at DateTime64(6, 'UTC') NOT NULL,
```

Comment:
do we need this? Don't we already will have `_timestamp` by default?

Reply:
That is a specific column introduced by the Kafka table if I'm not mistaken. I would argue that we should provide a schema for this table that makes sense and is explicit rather than relying on these virtual columns.

This is also not the merge_at time but the time that the message was sent to Kafka by the producer, which is semantically different.

However, I'm fine to remove this merged_at column if we don't think it is useful. It _may_ be useful and is something that might be annoying to backport.

Reply:
should we add this to the postgres table too then?

Reply:
I think it makes sense to record this yes. Maybe it's overkill by it's one of these things that you wished you had from the start but is annoying to add retroactively.

Reply:
@hazzadous is totally right here - `_timestamp` is the time the message was produced to kafka. We should only use `_timestamp` for analyzing the pipeline itself (for instance looking at the number of messages produced to a topic in a timeframe). We shouldn't use this field to infer anything with the data since it could be a retry or otherwise decoupled from the event's data itself.

Reply:
Re: Adding this field to Postgres: Question: Isn't Postgres written independently or before sending a message to Kafka? How would we know the value of `merged_at` to write it to Postgres too?

Reply:
Postgres would be written before I believe, and I think the plan is to write is at the point of merging two persons. I'm not sure what the value would be for the transitive case i.e. updating the values when there is a chain, but I guess you could still view that as the merge time of the whole chain.